content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1088 @@
+    * this is needed to check the condition in a thread-safe manner
+    * `wait` releases the lock
+```python
+cond.acquire()
+cond.wait()
+```
+* `notify`: wake up **one** thread waiting on the condition
+* `notify_all`: wake up **all** threads waiting on the condition
+
+----
+
+### Condition Variables - Demo
+
+```
+student@os:~/.../compute/lecture/demo/condition$ python3 condition.py
+Waiting thread: waiting for main thread...
+Main thread: type "notify" to notify waiting thread
+
+waiting thread is sleeping until woken up by main thread
+
+notify
+Waiting thread: notified by main thread
+```
+
+---
+
+### Thread-safety and Reentrancy
+
+* **Thread-safe function:** called simultaneously from multiple threads without race conditions
+* **Reentrant function:** preempted at any time and re-called from the same thread with the same effects
+    * Different from recursion:
+
+![Recursion vs reentrancy](./media/recursive-vs-reentrant.svg)
+
+----
+
+### Thread-safety and Reentrancy - Examples
+
+```c [1 - 8 | 10 - 16 | 18 - 24 | 26 - 30]
+/* Not thread-safe, not reentrant: */
+int tmp
+int add2(int x)
+{
+    tmp = x;
+    /* Preempt here */
+    return tmp + 2;
+}
+
+/* Reentrant, not thread-safe: */
+int tmp
+int add2(int x)
+{
+    tmp = x;
+    return a + 2;
+}
+
+/* Thread-safe not reentrant: */
+thread_local int tmp
+int add2(int x)
+{
+    tmp = x;
+    return tmp + 2;
+}
+
+/* Both thread-safe and reentrant: */
+int add2(int x)
+{
+    return x + 2;
+}
+```
+
+---
+
+## Cool Extra Stuff
+
+---
+
+### Create a new process - `fork` + `exec`
+
+* [Quiz](./quiz/exec-without-fork.md)
+* "Save" the shell before calling `exec("/bin/ls")`
+* Use `fork` and only the child calls `exec`
+
+![Fork and Exec Diagram](media/fork-exec.svg)
+
+----
+
+### Linux Syscalls
+
+```
+student@os:~/.../compute/lecture/demo/create-process$ strace ./fork_exec
+[...]
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7e83aa4810) = 5279
+[...]
+```
+* `fork` and `pthread_create` are `libc` wrappers to the same syscall: `clone`
+
+----
+
+### Libc vs Linux vs Windows
+
+* Libc approach - `system`:
+
+``` [1 | 2 | 3 | 5 | 7 | 9]
+student@os:~/.../compute/lecture/demo/create-process$ strace -ff -e clone,execve ./system
+execve("./system", ["./system"], 0x7ffff3f20008 /* 60 vars */) = 0
+clone(child_stack=0x7f3fa624aff0, flags=CLONE_VM|CLONE_VFORK|SIGCHLDstrace: Process 117770 attached
+ <unfinished ...>
+[pid 117770] execve("/bin/sh", ["sh", "-c", "ls -la"], 0x7ffdac7acfe8 /* 60 vars */ 
+[...]
+[pid 117770] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f9292b28850) = 117771
+strace: Process 117771 attached
+[pid 117771] execve("/usr/bin/ls", ["ls", "-la"], 0x55c700084048 /* 60 vars */) = 0
+[...]
+```
+
+* `system` first creates a new shell (`execve("/bin/sh"`)
+* This shell now spawns `ls -la` (`execve("/usr/bin/ls", ["ls", "-la"]`)
```

Comment:
The line is not properly displayed.

Reply:
I've placed both of these lines on another slide.